aws/cloudformation/bootstrap_chef_stack.sh.erb

```diff
@@ -126,15 +126,37 @@ cat <<JSON > $FIRST_BOOT
     "daemon": true
   },
 <% end -%>
-  "run_list": $RUN_LIST
+  "run_list": []
 }
 JSON
 
-# Provision via Chef.
-OPTIONS="<%=cookbooks_path ? '-z ' : ''%> -b $BRANCH -n $NODE_NAME -r '$RUN_LIST' -e $ENVIRONMENT -v $CHEF_VERSION"
+# Bootstrap Chef configuration: Install Chef, run chef-client with empty run-list to create node with first-boot config.
+OPTIONS="<%=cookbooks_path ? '-z ' : ''%> -b $BRANCH -n $NODE_NAME -r '' -e $ENVIRONMENT -v $CHEF_VERSION"
 sudo -u ubuntu bash -c "aws s3 cp <%=bootstrap_script_path%> - | sudo bash -s -- $OPTIONS"
 [ $? -eq 0 ] && STATUS=SUCCESS || STATUS=<%=daemon ? 'SUCCESS' : 'FAILURE'%>
 
+# Bash script that invokes chef-client with fetched config attributes and provided run-list.
+CHEF_CLIENT_BIN=<%=CHEF_CLIENT_BIN%>
```

Comment:
This looks fine to me, but I was curious on your thoughts on inlining this here, rather than uploading to S3 like we do for `chef-bootstrap.sh`. (Why isn't `chef-bootstrap.sh` just part of this script, anyway? Is it just to break up this script into smaller chunks?)

Reply:
> I was curious on your thoughts on inlining this here, rather than uploading to S3 like we do for `chef-bootstrap.sh`.

The parameterized arguments `${Chefconfig}` (which is a reference to the CFn resource) and `$RUN_LIST` (which is a reference to the `run_list` option passed from the stack ERB including this script) make it more complicated to render this script to S3 at compile-time like we do `chef-bootstrap.sh`, especially since this script is being invoked from the CI-build as well, not just from this bootstrap script.

> Why isn't chef-bootstrap.sh just part of this script, anyway? Is it just to break up this script into smaller chunks?

`chef-bootstrap.sh` was an experiment with deploying a self-contained bash script partly to break apart the bootstrap script as you guessed, but also because I thought that its purpose was narrow/isolated enough that it could be written as a standalone script, manually deployed once to a static location and touched very rarely after that.

That experiment turned out to be challenging as soon as it became a production-build dependency: deploying any meaningful change to the script required either careful backwards-compatibility with each change, and/or synchronization across all environments to prevent breaking the production workflow. So to avoid these issues (without setting up a whole more elaborate versioning, packaging and deployment system for such a tiny thing), I made a change to deploy it to a separate path per stack. So now it's basically tied to the stack anyway, contrary to the original design. It could be in-lined at this point with some additional work to reduce the overall complexity.